### ISSUE TITLE
Add support for custom craft types

### ISSUE THREAD

**Issue by 0gb-us** [minetest/minetest#1079]
Original Title: `Move minetest.register_craft() and minetest.get_craft_result() to Lua.`

Sapier asked me to submit this issue on SegFault22's behalf.

SegFault22 suggests that custom craft recipe types be definable in modules, using the same algorithms as the current recipe types. As discussed on IRC, this isn't feasible, in part because it does not allow custom algorithms to be used. However, as suggested by ShadowNinja, minetest.register_craft() and minetest.get_craft_result() could instead be moved from C++ to Lua. This would allow custom craft recipe types to reuse the algorithms used by the current recipe types, and would allow new recipe types to be seamlessly added by modules. This would also allow custom recipe types to define their own algorithms for more interesting craft results.

**Comment by rubenwardy** [comment]
> @MisterXtreme 
>
> I, and other modders would greatly appreciate custom recipe support.
> Q: Why?
> A: To allow a standard way for defining custom recipes, for example, right now to add an oven that gives you more bread when you put flour in it, you need hacky, non standard code to add a crafting recipe type for ovens, also meaning any crafting recipe guide mod needs me, or them to add support for my oven mod, for my recipes to even show up!
> 
> To tie into my last point, if a farming mod wants support for my crafting recipe system, it needs to specifically add support with a bunch of code, wheres with a custom recipe API, you could do minetest.getrecipetype(oven = 1), and you can detect my oven system, and use it.
> "But why not use is mod loaded?!" you ask?
> Simple, what if another mod implements my crafting system after I deprecate my oven mod, if it defines the same recipe type as mine, then the farming mod can just use it instead, allowing for a kind f recipe group.
> 
> Q: Why not use the code everyone else uses to add custom recipes?!"
> A: Please read the full post, that will help you understand as it answers quite a few concerns
> 
> TL;DR Read the whole thing, I can't condense it anymore then it already is.

**Comment by rubenwardy** [comment]
> @KristopherKram
> Custom types should be allowed in minetest.register_craft. Mods like technic and others have to work around this by creating custom code instead of easily creating a custom type. Example of a custom code work around is here: https://github.com/minetest-technic/technic/blob/master/technic/machines/register/grinder_recipes.lua .

**Comment by pauloue** [comment]
Any ideas how an API for this might look?

**Comment by SmallJoker** [comment]
```Lua
minetest.register_craft({
	type = "custom",
	recipe = {
		"wool:block",
		"dye:red",
	},
	output = function(player)
		return ItemStack("default:stick")
	end,
})
```

**Comment by pauloue** [comment]
You'd also need to be able to define the crafting method for use with `get_craft_result` and it should support custom fields like cooktime for e.g. an alloy furnace.
